### PR TITLE
Enhancement/mask

### LINKED ISSUE #472: When in a relation, user's sensitive data should be masked
By default, for privacy and security reasons, private user data should be masked when user is in a relation with another content-type.

**Behaviour:**
```
GET localhost:1337/custom

 {
        "_id": "...",
        "__v": 0,
        "owner": {
            "_id": "...",
            "username": "admin",
            "password": "...",
            "email": "...",
            "provider": "...",
            "role": 0,
            "__v": 0,
            "id": "..."
        },
        "id": "..."
    }

```

**Expected:**

```
GET localhost:1337/custom

 {
        "_id": "...",
        "__v": 0,
        "owner": {
            "_id": "...",
            "username": "admin",
            "__v": 0,
            "id": "..."
        },
        "id": "..."
    }

```

**In my opinion this should be the default case for any query on any user object, if the authenticated user itself is not requesting it.**

**Strapi version**: 3.0.0-alpha.8

### PR DESCRIPTION
**Pull request by Aurelsicoko**
Adding a new middleware `mask`. This middleware parses the successful JSON responses and tries to find a model pattern recursively. When there are matches, it removes from the response the model's attributes which have the `private` set to `true`.

Fix #472

### CODE REVIEW

File: packages/strapi/lib/middlewares/mask/index.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        const main = Object.keys(strapi.models).reduce((acc, current) => {
+          if (Object.values(strapi.models[current].attributes).find(attr => attr.private === true)) {
+            acc = true;
+          }
+
+          return acc;
+        }, false);
+
+        const plugins = Object.keys(strapi.plugins).reduce((acc, plugin) => {
+          acc = Object.keys(strapi.plugins[plugin].models).reduce((acc, model) => {
```
Comment (Aurelsicoko) [inline packages/strapi/lib/middlewares/mask/index.js]: There is a missing `if` condition. The latest plugin overrides previous value.

File: packages/strapi/lib/middlewares/mask/index.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+      if (enabled) {
+        strapi.app.use(async (ctx, next) => {
+          // Execute next middleware.
+          await next();
+
+          // Recursive to mask the private properties.
+          const mask = (payload) => {
+            if (_.isArray(payload)) {
+              return payload.map(value => mask(value));
```
Comment (Aurelsicoko) [inline packages/strapi/lib/middlewares/mask/index.js]:
```js
return payload.map(mask)
```

File: packages/strapi/lib/middlewares/mask/index.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+          matchs.push({
+            plugin,
+            model,
+            intersection
+          });
+        }
+      };
+
+      // Application models.
+      Object.keys(strapi.models).forEach(model => match(model));
```
Comment (Aurelsicoko) [inline packages/strapi/lib/middlewares/mask/index.js]:
```js
Object.keys(strapi.models).forEach(match);

...
```

**Comment by Aurelsicoko** [comment]
Missing documentation btw

**Comment by FullstackJack** [comment]
Looking forward to this in next release.

**Pull request merged**